spock-core/src/main/java/org/spockframework/runtime/extension/builtin/RetryIterationInterceptor.java

```diff
@@ -30,47 +30,48 @@
  * @since 1.2
  */
 public class RetryIterationInterceptor extends RetryBaseInterceptor implements IMethodInterceptor {
-  public RetryIterationInterceptor(Retry retry) {
+
+  private boolean finalIteration;
+  private final InnerRetryInterceptor retryInterceptor;
+  private List<Throwable> throwables;
+
+  public RetryIterationInterceptor(Retry retry, MethodInfo featureMethod) {
     super(retry);
+    this.retryInterceptor = new InnerRetryInterceptor(retry, condition);
+    featureMethod.addInterceptor(retryInterceptor);
   }
 
   @Override
   public void intercept(IMethodInvocation invocation) throws Throwable {
-    List<Throwable> throwableList = new ArrayList<>();
-      Queue<Throwable> throwables = new ConcurrentLinkedQueue<>();
+    throwables = new ArrayList<>();
```

Comment:
This is not really thread safe and will cause problems when used in parallel execution. We could either use a thread local, or force same thread execution for features with `@Retry`.

You can force parallel execution for embedded tests with this snippet in the given block:

```groovy
    runner.configurationScript {
      runner {
        parallel {
          enabled true
          fixed(4)
        }
      }
    }
```
When using parallel execution this tests becomes flaky: `org.spockframework.smoke.extension.RetryFeatureExtensionSpec#@Retry mode SETUP_FEATURE_CLEANUP executes setup and cleanup 12 times for @Unroll'ed feature`

Reply:
Thanks! Wasn't aware of the parallel configuration.

I hid the state behind ThreadLocals now which seems to work.

Reply:
Looks good so far, but I see that you didn't update the tests to enable parallel execution to check this.